### PR TITLE
mon/MDSMonitor: do not propose on error in prepare_update

### PR DESCRIPTION
**Pull request by batrick**
There is nothing to propose in pending map when an error is encountered.

Fixes: https://tracker.ceph.com/issues/58971

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by vshankar — COMMENTED** [review]
LGTM

File: src/mon/MDSMonitor.cc [hunk 1]
```diff
@@ -554,7 +554,7 @@ bool MDSMonitor::prepare_update(MonOpRequestRef op)
     } catch (const bad_cmd_get& e) {
       bufferlist bl;
       mon.reply_command(op, -EINVAL, e.what(), bl, get_last_committed());
-      return true;
+      return false; /* nothing to propose */
```
Comment (gregsfortytwo) [inline src/mon/MDSMonitor.cc:557]: The return value for prepare_update is true if the update has been *handled*. It has nothing to do with whether the paxos system needs to do a propose or not.

So this change is definitely bad.

File: src/mon/MDSMonitor.cc [hunk 2]
```diff
@@ -564,7 +564,7 @@ bool MDSMonitor::prepare_update(MonOpRequestRef op)
     ceph_abort();
   }
 
-  return true;
+  return false; /* nothing to propose! */
```
Comment (gregsfortytwo) [inline src/mon/MDSMonitor.cc:567]: Same as above, bad change.

**Review by gregsfortytwo — CHANGES_REQUESTED** [review]
As formatted this will break the monitor with unhandled requests.

I don't remember how we set up our internal interfaces for stuff like prepare_beacon(), but the interface for prepare_update() is very specific. You should all understand it if you're reviewing or making monitor changes @rishabh-d-dave @vshankar @batrick .

File: src/mon/MDSMonitor.cc (hunk 1 above)
Comment (gregsfortytwo) [reply src/mon/MDSMonitor.cc:557]: Cc @rishabh-d-dave as we discussed this verbally.

Glad the header docs are getting updated

**Comment by batrick** [comment]
jenkins test windows

**Comment by ljflores** [comment]
@batrick can you take a look at tracker no. 5 (https://tracker.ceph.com/issues/59380) in this list? I don't think it's related to your PR since I saw it in another testing branch as well. But see what you think since it involves an offline filesystem.

Otherwise, I didn't see anything related. If that tracker is unrelated in your opinion, feel free to merge.

Rados suite review: https://pulpito.ceph.com/?branch=wip-yuri2-testing-2023-03-30-0826

Failures:
    1. https://tracker.ceph.com/issues/51964
    2. https://tracker.ceph.com/issues/58946
    3. https://tracker.ceph.com/issues/58758
    4. https://tracker.ceph.com/issues/58585
    5. https://tracker.ceph.com/issues/59380 -- new tracker
    6. https://tracker.ceph.com/issues/59080
    7. https://tracker.ceph.com/issues/59057
    8. https://tracker.ceph.com/issues/59196

Details:
    1. qa: test_cephfs_mirror_restart_sync_on_blocklist failure - Ceph - CephFS
    2. cephadm: KeyError: 'osdspec_affinity' - Ceph - Orchestrator
    3. qa: fix testcase 'test_cluster_set_user_config_with_non_existing_clusterid' - Ceph - CephFS
    4. rook: failed to pull kubelet image - Ceph - Orchestrator
    5. rados/singleton-nomsgr: test failing from "Health check failed: 1 full osd(s) (OSD_FULL)" and "Health check failed: 1 filesystem is offline (MDS_ALL_DOWN)" - Ceph - RADOS
    6. mclock-config.sh: TEST_profile_disallow_builtin_params_modify fails when $res == $opt_val_new - Ceph - RADOS
    7. rados/test_envlibrados_for_rocksdb.sh: No rule to make target 'rocksdb_env_librados_test' on centos 8 - Ceph - RADOS
    8. ceph_test_lazy_omap_stats segfault while waiting for active+clean - Ceph - RADOS

**Comment by batrick** [comment]
> @batrick can you take a look at tracker no. 5 (https://tracker.ceph.com/issues/59380) in this list? I don't think it's related to your PR since I saw it in another testing branch as well. But see what you think since it involves an offline filesystem.
> 
> Otherwise, I didn't see anything related. If that tracker is unrelated in your opinion, feel free to merge.
> 
> Rados suite review: https://pulpito.ceph.com/?branch=wip-yuri2-testing-2023-03-30-0826
> 
> Failures: 1. https://tracker.ceph.com/issues/51964 2. https://tracker.ceph.com/issues/58946 3. https://tracker.ceph.com/issues/58758 4. https://tracker.ceph.com/issues/58585 5. https://tracker.ceph.com/issues/59380 -- new tracker 6. https://tracker.ceph.com/issues/59080 7. https://tracker.ceph.com/issues/59057 8. https://tracker.ceph.com/issues/59196
> 
> Details: 1. qa: test_cephfs_mirror_restart_sync_on_blocklist failure - Ceph - CephFS 2. cephadm: KeyError: 'osdspec_affinity' - Ceph - Orchestrator 3. qa: fix testcase 'test_cluster_set_user_config_with_non_existing_clusterid' - Ceph - CephFS 4. rook: failed to pull kubelet image - Ceph - Orchestrator 5. rados/singleton-nomsgr: test failing from "Health check failed: 1 full osd(s) (OSD_FULL)" and "Health check failed: 1 filesystem is offline (MDS_ALL_DOWN)" - Ceph - RADOS 6. mclock-config.sh: TEST_profile_disallow_builtin_params_modify fails when $res == $opt_val_new - Ceph - RADOS 7. rados/test_envlibrados_for_rocksdb.sh: No rule to make target 'rocksdb_env_librados_test' on centos 8 - Ceph - RADOS 8. ceph_test_lazy_omap_stats segfault while waiting for active+clean - Ceph - RADOS

Indeed, it doesn't look related and it seems other instances were found without this PR.

Will merge when my fs suite run is finished being analyzed.

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Main#24-May-2023

**Pull request merged**